### PR TITLE
Handle the deprecated port field in backend service.

### LINKED ISSUE #1715: IAP scanner failed when the backend service doesn't have a port
> [forseti-security] google.cloud.forseti.scanner.scanner(run): Error running scanner: IapScanner Traceback (most recent call last): File "/usr/local/lib/python2.7/dist-packages/forseti_security-2.0.0-py2.7.egg/google/cloud/forseti/scanner/scanner.py", line 112, in run scanner.run() File "/usr/local/lib/python2.7/dist-packages/forseti_security-2.0.0-py2.7.egg/google/cloud/forseti/scanner/scanners/iap_scanner.py", line 609, in run all_violations, _ = self._find_violations(iap_data) File "/usr/local/lib/python2.7/dist-packages/forseti_security-2.0.0-py2.7.egg/google/cloud/forseti/scanner/scanners/iap_scanner.py", line 594, in _find_violations for (iap_resources, project_resource_counts) in iap_data: File "/usr/local/lib/python2.7/dist-packages/forseti_security-2.0.0-py2.7.egg/google/cloud/forseti/scanner/scanners/iap_scanner.py", line 577, in _retrieve run_data.make_iap_resource(backend, parent.full_name)) File "/usr/local/lib/python2.7/dist-packages/forseti_security-2.0.0-py2.7.egg/google/cloud/forseti/scanner/scanners/iap_scanner.py", line 301, in make_iap_resource if self.is_alternate_service(backend_service, backend_service2): File "/usr/local/lib/python2.7/dist-packages/forseti_security-2.0.0-py2.7.egg/google/cloud/forseti/scanner/scanners/iap_scanner.py", line 338, in is_alternate_service backend_service2, instance_group2) File "/usr/local/lib/python2.7/dist-packages/forseti_security-2.0.0-py2.7.egg/google/cloud/forseti/scanner/scanners/iap_scanner.py", line 114, in instance_group_network_port port = int(backend_service.port) TypeError: int() argument must be a string or a number, not 'NoneType'

Need to handle the case when backend_service.port = None

### PR DESCRIPTION
**Pull request by joecheuk**
Resolves #1715

### CODE REVIEW

File: google/cloud/forseti/scanner/audit/blacklist_rules_engine.py [hunk 1]
```diff
@@ -220,6 +220,12 @@ def find_violations(self, instance_network_interface):
             project = network_and_project.group(1)
             network = network_and_project.group(2)
 
+            if not network_interface.access_configs:
+                LOGGER.warn('Network interface: %s, doesn\'t '
```
Comment (blueandgold) [inline google/cloud/forseti/scanner/audit/blacklist_rules_engine.py]: I would like this message to show that we are unable to find the violation here.  So, something like:
```
Unable to determine blacklist violation for network interface: %s, because it doesn't have access_configs.
```

File: google/cloud/forseti/scanner/scanners/iap_scanner.py [hunk 2]
```diff
@@ -272,7 +272,7 @@ def make_iap_resource(self, backend_service, project_full_name):
         for backend in backend_service.backends:
             instance_group = self.find_instance_group_by_url(
                 backend.get('group'))
-            if not instance_group:
+            if not instance_group or not backend_service.port:
```
Comment (blueandgold) [inline google/cloud/forseti/scanner/scanners/iap_scanner.py]: Are we still able to scan the resources correctly, if we suppress the missing port here?

i.e. will IapResource still be constructed correctly, and then evaluated later for violation?

File: google/cloud/forseti/scanner/audit/blacklist_rules_engine.py (hunk 1 above)
Comment (joecheuk) [reply google/cloud/forseti/scanner/audit/blacklist_rules_engine.py]: Updated

File: google/cloud/forseti/scanner/scanners/iap_scanner.py (hunk 2 above)
Comment (joecheuk) [reply google/cloud/forseti/scanner/scanners/iap_scanner.py]: Updated with detail comment.

File: google/cloud/forseti/scanner/scanners/iap_scanner.py [hunk 3]
```diff
@@ -111,12 +111,27 @@ def instance_group_network_port(backend_service, instance_group):
...
-                    break
+
+        # Field 'port' from backend service has been deprecated in favor of
+        # portName. PortName is required when the load balancing scheme is
+        # EXTERNAL. When the load balancing scheme is INTERNAL, this field
+        # is not used, it has the same behavior of port so we can just use
+        # portName to get the port from instance group.
+
+        if not backend_service.port_name:
+            # If backend_service.port_name is None, it's INTERNAL.
```
Comment (blueandgold) [inline google/cloud/forseti/scanner/scanners/iap_scanner.py]: nit: can you move this comment to above the `if` statement?

```
# If backend_service.port_name is None, it's INTERNAL.
if not backend_service.port_name:
    return None
```

File: google/cloud/forseti/scanner/scanners/iap_scanner.py [hunk 4]
```diff
@@ -111,12 +111,27 @@ def instance_group_network_port(backend_service, instance_group):
...
+            return None
+
+        port = None
+        for named_port in instance_group.named_ports or []:
+            if named_port.get('name') == backend_service.port_name:
+                port = int(named_port.get('port'))
+                break
+        if not port:
+            LOGGER.error('Unable to find backend_service.portName in '
+                         'instance_group.named_ports, NetworkPort cannot '
```
Comment (blueandgold) [inline google/cloud/forseti/scanner/scanners/iap_scanner.py]: The comma should be a period.

File: google/cloud/forseti/scanner/scanners/iap_scanner.py [hunk 5]
```diff
@@ -111,12 +111,27 @@ def instance_group_network_port(backend_service, instance_group):
...
+        # portName. PortName is required when the load balancing scheme is
+        # EXTERNAL. When the load balancing scheme is INTERNAL, this field
+        # is not used, it has the same behavior of port so we can just use
+        # portName to get the port from instance group.
+
+        if not backend_service.port_name:
+            # If backend_service.port_name is None, it's INTERNAL.
+            return None
+
+        port = None
```
Comment (blueandgold) [inline google/cloud/forseti/scanner/scanners/iap_scanner.py]: At this point, since we do have a valid backend service, do we not want to instantiate port as:
```
port = int(backend_service.port)
```

File: google/cloud/forseti/scanner/scanners/iap_scanner.py [hunk 6]
```diff
@@ -111,12 +111,27 @@ def instance_group_network_port(backend_service, instance_group):
...
+        for named_port in instance_group.named_ports or []:
+            if named_port.get('name') == backend_service.port_name:
+                port = int(named_port.get('port'))
+                break
+        if not port:
```
Comment (blueandgold) [inline google/cloud/forseti/scanner/scanners/iap_scanner.py]: If the comment above is valid, then we should not check for null port here.

File: google/cloud/forseti/scanner/scanners/iap_scanner.py (hunk 3 above)
Comment (joecheuk) [reply google/cloud/forseti/scanner/scanners/iap_scanner.py]: Updated.

File: google/cloud/forseti/scanner/scanners/iap_scanner.py (hunk 5 above)
Comment (joecheuk) [reply google/cloud/forseti/scanner/scanners/iap_scanner.py]: The field port is deprecated, we shouldn't be using it in our code.

You can find more details here:
https://cloud.google.com/compute/docs/reference/rest/v1/backendServices

File: google/cloud/forseti/scanner/scanners/iap_scanner.py (hunk 4 above)
Comment (joecheuk) [reply google/cloud/forseti/scanner/scanners/iap_scanner.py]: Updated.

File: google/cloud/forseti/scanner/scanners/iap_scanner.py (hunk 5 above)
Comment (blueandgold) [reply google/cloud/forseti/scanner/scanners/iap_scanner.py]: Could there be any legacy issue?  i.e. some customer who are configured this from before and have not migrated to the new config / field?

File: google/cloud/forseti/scanner/scanners/iap_scanner.py [hunk 7]
```diff
@@ -111,12 +111,31 @@ def instance_group_network_port(backend_service, instance_group):
...
+        # is not used, it has the same behavior of port so we can just use
+        # portName to get the port from instance group.
+
+        if not backend_service.port_name:
+            # Backend_service.port_name is None, it's INTERNAL.
+            return None
+
+        port = -1
+
+        if backend_service.port:
```
Comment (blueandgold) [inline google/cloud/forseti/scanner/scanners/iap_scanner.py]: Can you add a comment here to say that although deprecated, it's still returned by the API and might contain legacy data for customers who have not migrated.

File: google/cloud/forseti/scanner/scanners/iap_scanner.py [hunk 8]
```diff
@@ -329,14 +350,17 @@ def is_alternate_service(self, backend_service, backend_service2):
...
+
             for backend2 in backend_service2.backends:
                 instance_group2 = self.find_instance_group_by_url(
                     backend2.get('group'))
                 if not instance_group2:
                     continue
                 network_port2 = self.instance_group_network_port(
                     backend_service2, instance_group2)
-                if network_port != network_port2:
+                if not network_port2 or network_port != network_port2:
```
Comment (blueandgold) [inline google/cloud/forseti/scanner/scanners/iap_scanner.py]: Can you please break up the if conditions here into 2?  The code here are all standalone if's, and to me is also more clear.
```
if not network_port2:
    continue

if network_port != network_port2:
    continue

File: tests/scanner/iap_scanner_test.py [hunk 9]
```diff
@@ -510,16 +514,7 @@ def test_retrieve_resources(self):
             iap_scanner.IapResource(
                 project_full_name='organization/12345/project/foo/',
                 backend_service=BACKEND_SERVICES['bs1'],
-                alternate_services=set([
```
Comment (blueandgold) [inline tests/scanner/iap_scanner_test.py]: Why is this removed?

File: google/cloud/forseti/scanner/scanners/iap_scanner.py (hunk 7 above)
Comment (joecheuk) [reply google/cloud/forseti/scanner/scanners/iap_scanner.py]: Updated.

File: google/cloud/forseti/scanner/scanners/iap_scanner.py [hunk 10]
```diff
@@ -111,12 +111,29 @@ def instance_group_network_port(backend_service, instance_group):
...
+            # contain legacy data for customers who have not migrated.
+            port = int(backend_service.port)
+
         if backend_service.port_name:
             for named_port in instance_group.named_ports or []:
                 if named_port.get('name') == backend_service.port_name:
                     port = int(named_port.get('port'))
                     break
+        if port == -1:
+            LOGGER.error('Unable to find backend_service.portName in '
```
[the remark this answers is not in the document]
Comment (joecheuk) [reply google/cloud/forseti/scanner/scanners/iap_scanner.py]: Updated.
